{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import os\n",
    "import sys\n",
    "sys.path.append('/home/graduates/cyn/Cell2Map')\n",
    "from Cell2Map import common\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "\n",
    "from Cell2Map import map_utils\n",
    "from Cell2Map import autoencoder\n",
    "from Cell2Map import utils\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Noise level 5%, mean of 5 cells per spot\n",
    "## To use different simulation data, just change the data path\n",
    "scRNA_path=\"../data/Hippocampus/scRNA/scRNA_Noise05.csv\"\n",
    "cell_type_path=\"../data/Hippocampus/scRNA/cell_labels.csv\"\n",
    "st_path=\"../data/Hippocampus/5/ST_expression.csv\"\n",
    "coordinates_path=\"../data/Hippocampus/5/Coordinates.csv\"\n",
    "ground_truth_path=\"../data/Hippocampus/5/ground_truth.csv\"\n",
    "ground_truth = common.read_file(ground_truth_path)\n",
    "output_folder=\"../result/Hippocampus/xiaorong/distance_term\"\n",
    "mean_cell_numbers=5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_col='CellType'\n",
    "\n",
    "# Read the data from the specified paths\n",
    "# sc_adata: single-cell RNA-seq data\n",
    "# spatial_adata: spatial transcriptomics data\n",
    "# cell_type_number_eachspot_data: estimated number of each cell type in each spot\n",
    "# coordinates_data: coordinates of each spot\n",
    "sc_adata, spatial_adata, coordinates_data =\\\n",
    "        common.read_data(scRNA_path, cell_type_path, st_path, coordinates_path,celltype_col)\n",
    "\n",
    "# Extract the cell type information from the single-cell data\n",
    "# Rename the index and the 'CellType' column for clarity\n",
    "cell_type=pd.DataFrame(sc_adata.obs['CellType'])\n",
    "cell_type.index = [str(idx)[5:] for idx in cell_type.index]\n",
    "cell_type['CellType'] = [str(idx)[5:] for idx in cell_type['CellType']]\n",
    "\n",
    "# estimate the number of cells in each spot\n",
    "cell_number_to_node_assignment=common.estimate_cell_number_RNA_reads(spatial_adata.to_df().T, int(mean_cell_numbers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/graduates/anaconda3/envs/celloc/lib/python3.9/site-packages/scanpy/preprocessing/_normalization.py:207: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    }
   ],
   "source": [
    "sc_adata,spatial_adata=map_utils.process_adatas(sc_adata,spatial_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 23372 × 20419\n",
       "    obs: 'CellType'\n",
       "    var: 'n_cells'\n",
       "    uns: 'log1p', 'pca'\n",
       "    obsm: 'X_pca'\n",
       "    varm: 'PCs'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4656 × 20419\n",
       "    obs: 'rna_count_based_density'\n",
       "    var: 'n_cells'\n",
       "    uns: 'log1p'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spatial_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The graph contains 196250 edges, 23372 cells.\n",
      "8.396799589252096 neighbors per cell on average.\n",
      "The graph contains 163805 edges, 4656 cells.\n",
      "35.181486254295535 neighbors per cell on average.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 500/500 [00:55<00:00,  8.94it/s]\n"
     ]
    }
   ],
   "source": [
    "sc_adata,spatial_adata=autoencoder.embedding_feature(sc_adata,spatial_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/graduates/cyn/DEEPIAC/DEEPIAC/map_optimizer.py:49: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  self.d=torch.tensor(d,device=device, dtype=torch.float32)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seed:1000\n",
      "Total_loss: 20.757, gv_term: 0.903, vg_term: 0.490, density_term: 8.446, expression_term: 13.704, \n",
      "Total_loss: 20.722, gv_term: 0.917, vg_term: 0.490, density_term: 8.446, expression_term: 13.683, \n",
      "Total_loss: 20.695, gv_term: 0.923, vg_term: 0.490, density_term: 8.447, expression_term: 13.661, \n",
      "Total_loss: 20.671, gv_term: 0.926, vg_term: 0.490, density_term: 8.448, expression_term: 13.640, \n",
      "Total_loss: 20.649, gv_term: 0.928, vg_term: 0.490, density_term: 8.449, expression_term: 13.618, \n",
      "Total_loss: 20.627, gv_term: 0.930, vg_term: 0.490, density_term: 8.451, expression_term: 13.596, \n",
      "Total_loss: 20.606, gv_term: 0.931, vg_term: 0.490, density_term: 8.453, expression_term: 13.574, \n",
      "Total_loss: 20.586, gv_term: 0.931, vg_term: 0.490, density_term: 8.455, expression_term: 13.552, \n",
      "Total_loss: 20.566, gv_term: 0.930, vg_term: 0.490, density_term: 8.457, expression_term: 13.530, \n",
      "Total_loss: 20.547, gv_term: 0.930, vg_term: 0.490, density_term: 8.460, expression_term: 13.507, \n"
     ]
    }
   ],
   "source": [
    "Deep_map=map_utils.map_cell_to_space(sc_adata,spatial_adata,b_init=cell_number_to_node_assignment,device='cuda:0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 23372 × 4656\n",
       "    obs: 'CellType'\n",
       "    var: 'rna_count_based_density'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Deep_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_sp_map_df = pd.DataFrame(Deep_map.X, index=sc_adata.obs_names, columns=spatial_adata.obs_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1698436/3400311695.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  max_column_names.iloc[i,1]=ground_truth.loc[cell_id][0]\n",
      "/tmp/ipykernel_1698436/3400311695.py:33: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  max_column_names.iloc[i,2]=ground_truth.loc[cell_id][1]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Predict</th>\n",
       "      <th>Lable</th>\n",
       "      <th>CellType</th>\n",
       "      <th>predict_x</th>\n",
       "      <th>predict_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Values</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>V10882</th>\n",
       "      <td>S1</td>\n",
       "      <td>S1</td>\n",
       "      <td>Microglia</td>\n",
       "      <td>2</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V552</th>\n",
       "      <td>S2</td>\n",
       "      <td>S2</td>\n",
       "      <td>Cajal Retzius</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V864</th>\n",
       "      <td>S3</td>\n",
       "      <td>S3</td>\n",
       "      <td>Polydendrocyte</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V21370</th>\n",
       "      <td>S3</td>\n",
       "      <td>S3</td>\n",
       "      <td>Entorhinal</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V17277</th>\n",
       "      <td>S3</td>\n",
       "      <td>S3843</td>\n",
       "      <td>Polydendrocyte</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V17357</th>\n",
       "      <td>S4655</td>\n",
       "      <td>S4655</td>\n",
       "      <td>Polydendrocyte</td>\n",
       "      <td>90</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V15965</th>\n",
       "      <td>S4656</td>\n",
       "      <td>S4656</td>\n",
       "      <td>Entorhinal</td>\n",
       "      <td>90</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V709</th>\n",
       "      <td>S4656</td>\n",
       "      <td>S4656</td>\n",
       "      <td>Cajal Retzius</td>\n",
       "      <td>90</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V1649</th>\n",
       "      <td>S4656</td>\n",
       "      <td>S1953</td>\n",
       "      <td>Entorhinal</td>\n",
       "      <td>90</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V15392</th>\n",
       "      <td>S4656</td>\n",
       "      <td>S716</td>\n",
       "      <td>Entorhinal</td>\n",
       "      <td>90</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21022 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Predict  Lable        CellType  predict_x  predict_y\n",
       "Values                                                     \n",
       "V10882      S1     S1       Microglia          2         33\n",
       "V552        S2     S2   Cajal Retzius          3          7\n",
       "V864        S3     S3  Polydendrocyte          3         10\n",
       "V21370      S3     S3      Entorhinal          3         10\n",
       "V17277      S3  S3843  Polydendrocyte          3         10\n",
       "...        ...    ...             ...        ...        ...\n",
       "V17357   S4655  S4655  Polydendrocyte         90         29\n",
       "V15965   S4656  S4656      Entorhinal         90         31\n",
       "V709     S4656  S4656   Cajal Retzius         90         31\n",
       "V1649    S4656  S1953      Entorhinal         90         31\n",
       "V15392   S4656   S716      Entorhinal         90         31\n",
       "\n",
       "[21022 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove the first 5 characters from the column and index names of the DataFrame\n",
    "sc_sp_map_df.columns = [str(col)[5:] for col in sc_sp_map_df.columns]\n",
    "sc_sp_map_df.index = [str(idx)[5:] for idx in sc_sp_map_df.index]\n",
    "\n",
    "# Initialize an empty dictionary to store the results\n",
    "result_dict = {}\n",
    "\n",
    "# For each column in the DataFrame, find the cell names with the highest values\n",
    "# The number of cell names to find is determined by cell_number_to_node_assignment\n",
    "# Store the results in the result_dict dictionary\n",
    "for i, column in enumerate(sc_sp_map_df.columns.values):\n",
    "    result_dict[column] = sc_sp_map_df[column].nlargest(cell_number_to_node_assignment[i]).index\n",
    "\n",
    "# Convert the result_dict dictionary to a DataFrame\n",
    "# The 'Predict' column contains the spot names, and the 'Values' column contains the cell names\n",
    "# The 'Values' column is exploded so that each cell name has its own row\n",
    "max_column_names = pd.DataFrame({'Predict': list(result_dict.keys()), 'Values': [list(index_obj) for index_obj in result_dict.values()]}).explode('Values')\n",
    "\n",
    "# Convert the cell names in the 'Values' column to strings\n",
    "max_column_names['Values'] = max_column_names['Values'].astype(str)\n",
    "\n",
    "# Set the 'Values' column as the index of the DataFrame\n",
    "max_column_names = max_column_names.set_index('Values')\n",
    "max_column_names=max_column_names.assign(Lable='S0')\n",
    "max_column_names=max_column_names.assign(CellType='type')\n",
    "\n",
    "# Remove rows where the index is \"nan\"\n",
    "max_column_names = max_column_names[max_column_names.index!=\"nan\"]\n",
    "\n",
    "# For each row in the DataFrame, replace 'type' in the 'CellType' column with the actual cell type from the cell_type DataFrame\n",
    "for i,cell_id in enumerate(max_column_names.index.values):\n",
    "    max_column_names.iloc[i,1]=ground_truth.loc[cell_id][0]\n",
    "    max_column_names.iloc[i,2]=ground_truth.loc[cell_id][1]\n",
    "\n",
    "# Reindex the coordinates_data DataFrame based on the 'Predict' column of the max_column_names DataFrame\n",
    "# This gives the predicted coordinates for each cell\n",
    "predict_coordinates=coordinates_data.reindex(max_column_names['Predict'])\n",
    "\n",
    "# Add the predicted x and y coordinates to the max_column_names DataFrame\n",
    "max_column_names['predict_x']=predict_coordinates['row'].values\n",
    "max_column_names['predict_y']=predict_coordinates['col'].values\n",
    "\n",
    "max_column_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assigned_locations_path = os.path.join(output_folder, \"2\")\n",
    "os.makedirs(assigned_locations_path, exist_ok=True)\n",
    "result_csv_path=os.path.join(assigned_locations_path,\"result.csv\")\n",
    "max_column_names.to_csv(result_csv_path)\n",
    "scRNA_path=os.path.join(assigned_locations_path,\"sc_adata.h5ad\")\n",
    "sc_adata.write_h5ad(scRNA_path)\n",
    "spatial_path=os.path.join(assigned_locations_path,\"sp_adata.h5ad\")\n",
    "spatial_adata.write_h5ad(spatial_path)\n",
    "deep_map_path=os.path.join(assigned_locations_path,\"deep_map.h5ad\")\n",
    "Deep_map.write_h5ad(deep_map_path)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "celloc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
