{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import random\n",
    "import sys\n",
    "sys.path.append('/home/graduates/cyn/Cell2Map')\n",
    "from Cell2Map import common\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import sys\n",
    "\n",
    "from Cell2Map import map_utils\n",
    "from Cell2Map import autoencoder\n",
    "from Cell2Map import utils\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "scRNA_path=\"../data/DCIS2/sc_RNA_DCIS2.h5ad\"\n",
    "cell_type_path=None\n",
    "st_path=\"../data/DCIS2/spatial_DCIS2.h5ad\"\n",
    "coordinates_path=None\n",
    "output_folder=\"../result/DCIS2/\"\n",
    "mean_cell_numbers=5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_col='cell_type'\n",
    "\n",
    "# Read the data from the specified paths\n",
    "# sc_adata: single-cell RNA-seq data\n",
    "# spatial_adata: spatial transcriptomics data\n",
    "# cell_type_number_eachspot_data: estimated number of each cell type in each spot\n",
    "# coordinates_data: coordinates of each spot\n",
    "sc_adata, spatial_adata, coordinates_data =\\\n",
    "        common.read_data(scRNA_path, cell_type_path, st_path, coordinates_path,celltype_col)\n",
    "\n",
    "# Extract the cell type information from the single-cell data\n",
    "# Rename the index and the 'CellType' column for clarity\n",
    "cell_type=pd.DataFrame(sc_adata.obs[celltype_col])\n",
    "cell_type.index = [str(idx)[5:] for idx in cell_type.index]\n",
    "cell_type['CellType'] = [str(idx)[5:] for idx in cell_type[celltype_col]]\n",
    "\n",
    "# estimate the number of cells in each spot\n",
    "cell_number_to_node_assignment=common.estimate_cell_number_RNA_reads(spatial_adata.to_df().T, int(mean_cell_numbers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/graduates/anaconda3/envs/celloc/lib/python3.9/site-packages/scanpy/preprocessing/_normalization.py:207: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    }
   ],
   "source": [
    "sc_adata,spatial_adata=map_utils.process_adatas(sc_adata,spatial_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 3587 × 19840\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.1', 'seurat_clusters', 'cell_type', 'CellType'\n",
      "    var: 'n_cells'\n",
      "    uns: 'log1p', 'pca'\n",
      "    obsm: 'X_pca'\n",
      "    varm: 'PCs'\n",
      "AnnData object with n_obs × n_vars = 1567 × 19840\n",
      "    obs: 'in_tissue', 'array_row', 'array_col', 'rna_count_based_density'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells'\n",
      "    uns: 'spatial', 'log1p'\n",
      "    obsm: 'spatial'\n"
     ]
    }
   ],
   "source": [
    "print(sc_adata)\n",
    "print(spatial_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The graph contains 26051 edges, 3587 cells.\n",
      "7.262614998606077 neighbors per cell on average.\n",
      "The graph contains 0 edges, 1567 cells.\n",
      "0.0 neighbors per cell on average.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 500/500 [00:10<00:00, 48.77it/s]\n"
     ]
    }
   ],
   "source": [
    "sc_adata,spatial_adata=autoencoder.embedding_feature(sc_adata,spatial_adata,k_cutoff=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/graduates/cyn/Cell2spotGATE/Cell2spotGATE/map_optimizer.py:50: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  self.d=torch.tensor(d,device=device, dtype=torch.float32)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seed:1000\n",
      "Total_loss: 12.339, gv_term: 1.373, vg_term: 0.729, density_term: 8.976, expression_term: 6.886, distance_term:1.421,\n",
      "Total_loss: 11.947, gv_term: 1.437, vg_term: 0.729, density_term: 9.003, expression_term: 6.520, distance_term:1.411,\n",
      "Total_loss: 11.648, gv_term: 1.437, vg_term: 0.729, density_term: 9.068, expression_term: 6.155, distance_term:1.410,\n",
      "Total_loss: 11.435, gv_term: 1.394, vg_term: 0.729, density_term: 9.156, expression_term: 5.814, distance_term:1.412,\n",
      "Total_loss: 11.288, gv_term: 1.332, vg_term: 0.729, density_term: 9.257, expression_term: 5.510, distance_term:1.418,\n",
      "Total_loss: 11.183, gv_term: 1.270, vg_term: 0.729, density_term: 9.364, expression_term: 5.246, distance_term:1.428,\n",
      "Total_loss: 11.101, gv_term: 1.212, vg_term: 0.730, density_term: 9.479, expression_term: 5.010, distance_term:1.448,\n",
      "Total_loss: 11.022, gv_term: 1.157, vg_term: 0.730, density_term: 9.614, expression_term: 4.785, distance_term:1.490,\n",
      "Total_loss: 10.936, gv_term: 1.105, vg_term: 0.730, density_term: 9.775, expression_term: 4.563, distance_term:1.567,\n",
      "Total_loss: 10.855, gv_term: 1.061, vg_term: 0.730, density_term: 9.935, expression_term: 4.371, distance_term:1.660,\n",
      "Total_loss: 10.793, gv_term: 1.027, vg_term: 0.731, density_term: 10.063, expression_term: 4.230, distance_term:1.741,\n",
      "Total_loss: 10.739, gv_term: 1.003, vg_term: 0.732, density_term: 10.138, expression_term: 4.140, distance_term:1.804,\n",
      "Total_loss: 10.677, gv_term: 0.984, vg_term: 0.734, density_term: 10.184, expression_term: 4.075, distance_term:1.863,\n",
      "Total_loss: 10.612, gv_term: 0.968, vg_term: 0.737, density_term: 10.224, expression_term: 4.012, distance_term:1.919,\n",
      "Total_loss: 10.552, gv_term: 0.950, vg_term: 0.739, density_term: 10.265, expression_term: 3.946, distance_term:1.971,\n",
      "Total_loss: 10.504, gv_term: 0.933, vg_term: 0.740, density_term: 10.297, expression_term: 3.892, distance_term:2.013,\n",
      "Total_loss: 10.473, gv_term: 0.921, vg_term: 0.741, density_term: 10.311, expression_term: 3.861, distance_term:2.037,\n",
      "Total_loss: 10.455, gv_term: 0.915, vg_term: 0.741, density_term: 10.314, expression_term: 3.846, distance_term:2.049,\n"
     ]
    }
   ],
   "source": [
    "Deep_map=map_utils.map_cell_to_space(sc_adata,spatial_adata,\n",
    "learning_rate=0.005,\n",
    "    num_epochs=1800,\n",
    "    lambda_d=1.22,\n",
    "    lambda_g1=1.75,\n",
    "    lambda_g2=0.78,\n",
    "    b_init=cell_number_to_node_assignment,\n",
    "    alpha=2.05,\n",
    "    lambda_mahalanobis=0.65,\n",
    "    lambda_distance=0.012,\n",
    "    device='cuda:0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_sp_map_df = pd.DataFrame(Deep_map.X, index=sc_adata.obs_names, columns=spatial_adata.obs_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3477685/2128575498.py:31: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  max_column_names.iloc[i,-1]=cell_type.loc[cell_id][0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                             Predict    CellType  predict_x  predict_y\n",
      "Values                                                                \n",
      "GAATGAAAGGTAGCCA  AAACAAGTATCTCCCA-1  Nornal epi       2939       1190\n",
      "TTATGCTAGGGTCGAT  AAACAAGTATCTCCCA-1  Nornal epi       2939       1190\n",
      "GGATTACGTTTACTCT  AAACAAGTATCTCCCA-1      Clone1       2939       1190\n",
      "GTGCGGTGTCATACTG  AAACAAGTATCTCCCA-1  Nornal epi       2939       1190\n",
      "CGATGGCCAACGATGG  AAACAAGTATCTCCCA-1  Nornal epi       2939       1190\n",
      "...                              ...         ...        ...        ...\n",
      "TGGCTGGCAATAGAGT  TTGTTTGTGTAAATTC-1      Clone2        690       2693\n",
      "AGAGCTTGTCTAGGTT  TTGTTTGTGTAAATTC-1     Myeloid        690       2693\n",
      "TTCTTAGTCGGAGGTA  TTGTTTGTGTAAATTC-1     Myeloid        690       2693\n",
      "AACACGTGTAGCGTCC  TTGTTTGTGTAAATTC-1      Clone2        690       2693\n",
      "AGGCCGTAGCGATTCT  TTGTTTGTGTAAATTC-1      Clone1        690       2693\n",
      "\n",
      "[7059 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Remove the first 5 characters from the column and index names of the DataFrame\n",
    "sc_sp_map_df.columns = [str(col)[5:] for col in sc_sp_map_df.columns]\n",
    "sc_sp_map_df.index = [str(idx)[5:] for idx in sc_sp_map_df.index]\n",
    "\n",
    "# Initialize an empty dictionary to store the results\n",
    "result_dict = {}\n",
    "\n",
    "# For each column in the DataFrame, find the cell names with the highest values\n",
    "# The number of cell names to find is determined by cell_number_to_node_assignment\n",
    "# Store the results in the result_dict dictionary\n",
    "for i, column in enumerate(sc_sp_map_df.columns.values):\n",
    "    result_dict[column] = sc_sp_map_df[column].nlargest(cell_number_to_node_assignment[i]).index\n",
    "\n",
    "# Convert the result_dict dictionary to a DataFrame\n",
    "# The 'Predict' column contains the spot names, and the 'Values' column contains the cell names\n",
    "# The 'Values' column is exploded so that each cell name has its own row\n",
    "max_column_names = pd.DataFrame({'Predict': list(result_dict.keys()), 'Values': [list(index_obj) for index_obj in result_dict.values()]}).explode('Values')\n",
    "\n",
    "# Convert the cell names in the 'Values' column to strings\n",
    "max_column_names['Values'] = max_column_names['Values'].astype(str)\n",
    "\n",
    "# Set the 'Values' column as the index of the DataFrame\n",
    "max_column_names = max_column_names.set_index('Values')\n",
    "max_column_names=max_column_names.assign(CellType='type')\n",
    "\n",
    "# Remove rows where the index is \"nan\"\n",
    "max_column_names = max_column_names[max_column_names.index!=\"nan\"]\n",
    "\n",
    "# For each row in the DataFrame, replace 'type' in the 'CellType' column with the actual cell type from the cell_type DataFrame\n",
    "for i,cell_id in enumerate(max_column_names.index.values):\n",
    "    max_column_names.iloc[i,-1]=cell_type.loc[cell_id][0]\n",
    "\n",
    "# Reindex the coordinates_data DataFrame based on the 'Predict' column of the max_column_names DataFrame\n",
    "# This gives the predicted coordinates for each cell\n",
    "predict_coordinates=coordinates_data.reindex(max_column_names['Predict'])\n",
    "\n",
    "# Add the predicted x and y coordinates to the max_column_names DataFrame\n",
    "max_column_names['predict_x']=predict_coordinates['X'].values\n",
    "max_column_names['predict_y']=predict_coordinates['Y'].values\n",
    "\n",
    "# Print the DataFrame\n",
    "print(max_column_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assigned_locations_path = os.path.join(output_folder+\"Cell2Map_DCIS2.csv\")\n",
    "max_column_names.to_csv(assigned_locations_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "celloc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
